Query-frontend: accurate tracking of samples processed from cache

Konstantinov-Innokentii opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

#### What this PR does
This PR leverages perStepStats to accurately track samples processed from cache, without approximation. 
It's a follow-up for https://github.com/grafana/mimir/pull/11164 and https://github.com/grafana/mimir/pull/11416

#### How it works

- Added `GetStats() `and `SetStats()` MetricsReqiestInterface  and its implementations. It allows to manipulate a "stats" query param, which is required for MQE and Prometheus Query Engine to return PerStepStats.
- This param is always set in the [SplitAndCacheMiddleware](https://github.com/grafana/mimir/pull/11719/files#diff-655764486d839037085bd64ff1e3f08c12902660a2a083608ac67f55c07e3434R155), so each request will have a PerStepStats gathered.
- PerStepStats are [stored](https://github.com/grafana/mimir/pull/11719/files) in Extents instead of approximation.
- PerStepStats from cache are [merged](https://github.com/grafana/mimir/pull/11719/files#diff-655764486d839037085bd64ff1e3f08c12902660a2a083608ac67f55c07e3434R326) with the PerStepStats from a downstream request and sum is reported in the `cortex_query_samples_processed_cache_adjusted_total`
- Bunch of [integration tests ](https://github.com/grafana/mimir/pull/11719/files#diff-5d39b940451aba1e23d7e664ca574f63bbe521344e67670e9beefe9b082c348c)for that 


**Note**, that this functionality is under the `query-frontend.cache-samples-processed-stats` feature flag, so It's needed to enable this flag progressively for each release wave, to keep PerStepStats collection after the release. It also affects behaviour of the `cortex_query_samples_processed_cache_adjusted_total` metric. If `query-frontend.cache-samples-processed-stats` is not enabled if will report 0 samples.

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`. If changelog entry is not needed, please add the `changelog-not-needed` label to the PR.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

dimitarvdimitrov commented on cmd/mimir/config-descriptor.json:
do we need these to be configurable? is there a tradeoff for having them always enabled?

Konstantinov-Innokentii replied:
I mostly followed Prometheus and Cortex implementations here. My thoughts however, is that PerStepStats are not really needed by default for OSS or GEM users, so I put it behind feature flag.

dimitarvdimitrov replied:
isn't `querier.per-step-stats-enabled` also controlled by the query parameter - why also have a config option? Only `query-frontend.cache-samples-processed-stats` is something that's not controllable per-query and could use its own config option.

Konstantinov-Innokentii replied:
> isn't querier.per-step-stats-enabled also controlled by the query parameter - why also have a config option?

To report PerStep stats, Prometheus wants **both** config flag and the query parameter passed.
Under the hood, the `per-step-stats-enabled ` flag is passed down to the QueryEngine. When creating a new query, Prometheus reads the `r.Form("stats")`and if parameter is passed and engine has it's enabled too, PerStepStas will be reported.


In Mimir we reuse prometheus code which reads the `r.Form("stats")`, so we anyway need to pass it, but it's true that we have some options around the `per-step-stats-enabled` flag.
1. Introduce `querier.per-step-stats-enabled` 
2. hardcode `enablePerStepStas=true`when creating an Engine in Querier.

I find config more bit elegant - there is no hardcoded configs and it mimis Prometheus behaviour.

dimitarvdimitrov replied:
Right now, the reason to add a config option is for consistency in operating Mimir and operating Prometheus. We shouldn't be trying to have consistency there - they are two very different systems and have diverged a long time ago.

i'd be happy to add a config option, but we have to understand it first. If there is a practical reason to add the option, sure; otherwise, we're just piling up options, which even we can't recommend when to have on and when to have off.

Konstantinov-Innokentii replied:
If consistency here is not a concern, I can definitely adjust querier to always enable per step stats in engine and rely on a stats param only 👍🏻

dimitarvdimitrov commented on pkg/frontend/querymiddleware/codec.go:
```diff
@@ -138,6 +138,8 @@ type MetricsQueryRequest interface {
 	// GetLookbackDelta returns the lookback delta for the request.
 	GetLookbackDelta() time.Duration
+	// GetStats returns the stats parameter for the request.
+	GetStats() string
```
can we use an enum type here instead of a string?

Konstantinov-Innokentii replied:
Bit more details - here is some Prom code, which builds QueryOptions from request:
```
func extractQueryOpts(r *http.Request) (promql.QueryOpts, error) {
	var duration time.Duration

	if strDuration := r.FormValue("lookback_delta"); strDuration != "" {
		parsedDuration, err := parseDuration(strDuration)
		if err != nil {
			return nil, fmt.Errorf("error parsing lookback delta duration: %w", err)
		}
		duration = parsedDuration
	}

	return promql.NewPrometheusQueryOpts(r.FormValue("stats") == "all", duration), nil
}
```
and the `r.FormValue("stats") == "all"` goes to **enablePerStepStats** param of the `NewPrometheusQueryOpts`

We reuse this code in Querier, so that's why we have to pass "stats".

Konstantinov-Innokentii replied:
I did a pass on making it enum:
type StatsParam string

```
const (
	// StatsNone means no stats collection
	StatsNone StatsParam = ""
	// StatsAll means collect all stats including PerStepStats
	StatsAll StatsParam = "all"
	// StatsTrue means collect all stats except PerStepStats.
        // Prometheus doesn't specify concrete value, taking everything which is non empty string.
	StatsTrue StatsParam = "true" 
)

// String returns the string representation of StatsParam
func (s StatsParam) String() string {
	return string(s)
}
```
There is one caveat with this approach that I would like to hear what you think about. So, Prometheus doesn't specify values for the stat parameter.  If` stats != ""` it will return some stats in the Response and if `stats == "all"` it will return some stats + PerStepStats. It leads to some weird code like:
```
func ParseStatsParam(s string) StatsParam {
	switch s {
	case "all":
		return StatsAll
	case "":
		return StatsNone
	default:
		return StatsTrue
	}
}
```
Two things I don't really like with this are:
1. Some extra-behaviour of rewriting the value of stats param
2. If prometheus will ever change this behaviour, let's say will return stats on stats=true, instead of any non-empty value, this code will not work properly.

I don't have a strong point about what is better; neither option is ideal, so I'm happy to proceed with the option you prefer.

dimitarvdimitrov replied:
we agreed to keep them as separate concepts

Konstantinov-Innokentii replied:
We had a call and agreed that enum here will not help, since prometheus supports not only "all", but any non-empty string.

dimitarvdimitrov commented on pkg/frontend/querymiddleware/results_cache.go:
```diff
@@ -474,18 +473,24 @@ func newAccumulator(base Extent) (*accumulator, error) {
 		return Extent{}, err
 	}
+
+	extentPerStepStats := make([]StepStat, len(perStepStats))
+	for i, stat := range perStepStats {
+		extentPerStepStats[i] = StepStat(stat)
+	}
```
do you have any numbers on how much this would increase the cache size? should we try to delta-encode the per-step stats for example?

Konstantinov-Innokentii replied:
This PR will not increase the cache size; it changes the method of calculating cachedStats. However, they have been stored in a cache for ~ a month now after [this](https://github.com/grafana/mimir/pull/11164) was merged and released.
Speaking about numbers, it's protobuf encoded in cache, so I didn't do any real math around using delta-encoding.

Konstantinov-Innokentii replied:
I kept an eye on the Query results cache hit ratio hit rate when was testing it in dev cell and in ops after the rollout, (AFAIK it was released in ops on 04-24) and I didn't notice any sudden hit rate drops. Is there anything else we can check?

dimitarvdimitrov commented on pkg/frontend/querymiddleware/split_and_cache.go:
```diff
@@ -271,10 +284,11 @@ func (s *splitAndCacheMiddleware) Do(ctx context.Context, req MetricsQueryReques
 				downstreamRes := splitReq.downstreamResponses[downstreamIdx]
 				downstreamStats := splitReq.downstreamStatistics[downstreamIdx]
 				if !isResponseCachable(downstreamRes) {
+					fmt.Println("Response not cachable")
```
leftover debug statement; there's one above too

dimitarvdimitrov commented on pkg/querier/stats/stats.proto:
```diff
@@ -15,7 +15,7 @@ import "google/protobuf/duration.proto";
 option (gogoproto.unmarshaler_all) = true;
 
-message Stats {
+message StatsData {
```
this looks unnecessary

Konstantinov-Innokentii replied:
Well, I could use some help with that.  The story is that I need to add a mutex to make Stats.PerStepStats array concurrency-safe, especially when merging stats.
To achieve this, I [renamed](https://github.com/grafana/mimir/pull/11719/files#diff-8ca2404f80bab2e8d434ef41516c62c98fa2d642471ba0d8e29006c38eecb5deR18) the Protobuf message  StatsData and [wrapped](https://github.com/grafana/mimir/pull/11719/files#diff-25b9719633031df951dfb7a15155940c34efc0f34c2f8b18947bc8335fb3d465R46) it in a new Stats struct that includes a mutex.
However, frontend.proto uses the Stats message in the ClientToFrontend message, which is leading to some lint [errors](https://github.com/grafana/mimir/actions/runs/15630408459/job/44033201239?pr=11719#step:8:11).
The solution I'm considering is to rename StatsData back to Stats and introduce a new WrappedStats type that embeds Stats and includes a mutex. This way, the Protobuf-generated Stats would be used only when sending data between the querier and the frontends, while the rest of the code would interact with WrappedStats.
The downside is that I’d need to write code to convert between Stats and WrappedStats when encoding and decoding responses.

Konstantinov-Innokentii replied:
Thanks, I'll try that.

dimitarvdimitrov commented on pkg/frontend/querymiddleware/codec.go:
```diff
@@ -365,8 +369,9 @@ func (c prometheusCodec) decodeRangeQueryRequest(r *http.Request) (MetricsQueryR
 	var options Options
 	decodeOptions(r, &options)
 
+	stats := reqValues.Get("stats")
```
there is already an HTTP header which requests stats for the `server-timing` response header, can we reuse that? 

https://github.com/grafana/mimir/blob/8d8a0a82dc8b65222d389e0dfe8c291ec7869edb/pkg/frontend/transport/handler.go#L46

Konstantinov-Innokentii replied:
I'm not sure. The problem is that Prometheus Query Engine will Report per step stats under two conditions:
1. Feature flag is enabled
2. Request has a stats="all" query param.

Yes, we currently use MQE, but it might be the case that it's disabled in favour of Prometheus Query Engine, because of a bug or something, so that's why I had to implement what Prometheus wants.

dimitarvdimitrov replied:
i didn't know prometheus supports this too. So is your plan to also return these stats in the client response too?

It's unfortunate both projects implemented the same. Can we at least make it less confusing to use? I'm thinking that we can treat the header and the query param would mean the same thing - include `Server-Timing` response and also return per-step statistics.

dimitarvdimitrov replied:
in a DM we agreed to keep them as separate concepts because we always want prometheus stats for internal billing, but maybe not always want the header stats

dimitarvdimitrov commented on pkg/querier/stats/stats.pb.go:
```diff
@@ -54,19 +54,21 @@ type Stats struct {
 	// The number of subqueries that were spun off as actual range queries in order to execute the full query
 	SpunOffSubqueries uint32 `protobuf:"varint,12,opt,name=spun_off_subqueries,json=spunOffSubqueries,proto3" json:"spun_off_subqueries,omitempty"`
+	// SamplesProcessedPerStep represents the number of samples scanned per step timestamp while evaluating a query.
+	SamplesProcessedPerStep []StepStat `protobuf:"bytes,13,rep,name=samples_processed_per_step,json=samplesProcessedPerStep,proto3" json:"samples_processed_per_step"`
```
Have you considered having a different format for this? Right now if we wanted to add another type of statistics that we want to track per step, we would have to duplicate this array. So we'd have two arrays: one for SamplesProcessed and one for the next statistic. At least the timestamps would repeat.

WDYT about having single array where every element is a timestamp and a multiple values which have names. So in this case the field here would be called PerStepStatistics. And then the StepStats would be something like {Timestamp, SamplesProcessed}. (again - not sure if we need to store every timestamp or if start,end,step are enough)

Konstantinov-Innokentii replied:
We had a chat about it. Passing PerStepStats is in fact adding just adding one more series to the response, so we shouldn't be too concerned about it.

dimitarvdimitrov commented on pkg/querier/stats_renderer.go:
```diff
@@ -15,6 +14,23 @@ func StatsRenderer(ctx context.Context, s *promql_stats.Statistics, param string
 	}
-	return prom_api.DefaultStatsRenderer(ctx, s, param)
+	// Mimir doesn't use stats from Prometheus Response in favour of mimirStats above.
+	return nil
```
we still claim that our querier has a prometheus-compatible interface. If we want this to stay true, we have to keep these stats :(

At the same time, I see that we will transport the stats twice, which is not great, so I'm tempted to make this breaking change now and document it as caveat until 3.0. I added it to "2025-04 Mimir 3.0 Features and Changes" as a proposal where we can break this compatibility promise. Do you want to bring this up during a weekly call where we can discuss it?

Konstantinov-Innokentii replied:
I didn't know that; thanks for pointing it out. With that knowledge, I wonder if we should do the same as Cortex did—do not pass PerStep stats in the querier stats, but pass them in the PrometheusResponse and store in Extent response. Here is the [PR](https://github.com/cortexproject/cortex/pull/4708).
The idea behind not doing that was not to introduce a mess with some piece of stats passed in the queries stats and some in the Prometheus response, but passing them around twice seems like an even worse idea.

dimitarvdimitrov replied:
Taking them from the Prometheus response seems less surprising. Will this be hard to implement given the groundwork from https://github.com/grafana/mimir/pull/11164?

An alternative would be to strip the stats for internal responses. Here we make a differentiation between requests from the query-frontend/query-scheduler and requests coming directly to the querier. https://github.com/grafana/mimir/blob/699a122aba83da75ccef884d63af468da4f5127e/pkg/mimir/modules.go#L620-L666

But I'm not convinced we should be keeping this compatibility promise.

> what level of compatibility are we aiming to provide?

we're only looking at API-level compatibility - so the querier should feel like a Prometheus server to users/clients.

dimitarvdimitrov commented on pkg/frontend/querymiddleware/results_cache.go:
```diff
@@ -757,3 +738,111 @@ func mergeSamplesProcessedPerStep(a, b []StepStat) []StepStat {
+}
+
+// buildMinHeap builds a min-heap in-place
+func buildMinHeap(heap []heapItem) {
```
let's use the `container/heap` package instead of implementing our own heap.

Konstantinov-Innokentii replied:
Thanks

dimitarvdimitrov commented on pkg/frontend/querymiddleware/results_cache_test.go:
```diff
@@ -137,14 +135,14 @@ func mkExtentWithEvenPerStepSamplesProcessed(start, end int64, step int64, sampl
 
 func mxEvenlyDistributedExtentPerStepStats(start, end int64, step int64, samplesPerStep int64) []StepStat {
```
what does `mx` mean?

Konstantinov-Innokentii replied:
It's a typo, supposed to be "mk"

dimitarvdimitrov commented on pkg/frontend/querymiddleware/results_cache_test.go:
```diff
@@ -137,14 +135,14 @@ func mkExtentWithEvenPerStepSamplesProcessed(start, end int64, step int64, sampl
 func mxEvenlyDistributedExtentPerStepStats(start, end int64, step int64, samplesPerStep int64) []StepStat {
 	numSteps := int((end-start)/step) + 1
-	stats := make([]StepStat, numSteps)
+	s := make([]StepStat, numSteps)
```
nit: we don't need this change, right?

dimitarvdimitrov commented on pkg/frontend/querymiddleware/results_cache_test.go:
```diff
@@ -891,7 +944,8 @@ func TestFilterRecentCacheExtents(t *testing.T) {
 			// First extent - 1 hour with 1000ms step is a 3601 data points - 36010 samples - not truncated
-			// Second extent - 2 hours with 1000ms step is a 7201 data points - 72010 samples - truncated to 54010 samples
+			// Second extent - 2 hours with 1000ms step is a 7201 data points - 72010 samples.
+			// Truncated to 54010 samples.
```
nit but could be useful to say why it's truncated to 54010 exactly

Konstantinov-Innokentii replied:
Added more details, thanks

dimitarvdimitrov commented on pkg/frontend/querymiddleware/split_and_cache.go:
```diff
@@ -143,18 +146,26 @@ func (s *splitAndCacheMiddleware) Do(ctx context.Context, req MetricsQueryReques
+	cacheUnalignedRequests := validation.AllTrueBooleansPerTenant(tenantIDs, s.limits.ResultsCacheForUnalignedQueryEnabled)
+
+	// Force queier to track PerStepStats, so they could be cached.
+	if s.cacheSamplesProcessedStats {
```
i'm not sure if this shouldn't be called something like `computeSamplesProcessedStats` - it controls whether we compute the stats at all, not only whether we don't cache them. To me cache means that the thing we're caching already exists, we just choose to store it in the cache. 

nitpick i guess

Konstantinov-Innokentii replied:
I'm not fully happu with name, but let me give my perspective - w anyway computing Sample Processed, so it's more like `computeSamplesProcessedPerStep`. On the other hand, we compute them only to count SamplesProcessed from cache correctly, so that's why I ended up with cacheSamplesProcessedStats - to highlight feature, not the implementation

dimitarvdimitrov commented on pkg/frontend/querymiddleware/stats.go:
```diff
@@ -161,9 +161,9 @@ type QueryDetails struct {
 	ResultsCacheHitBytes  int
-	// SamplesProcessedFromCache represents the total number of samples processed by queriers to produce the result
-	// that has been stored in the query-frontend cache and then fetched to produce the current full query result.
-	SamplesProcessedFromCache uint64
```
there's a test which still references SamplesProcessedFromCache both in the name and in a comment

dimitarvdimitrov commented on pkg/frontend/querymiddleware/split_and_cache.go:
```diff
@@ -304,6 +313,28 @@ func (s *splitAndCacheMiddleware) Do(ctx context.Context, req MetricsQueryReques
 
+	samplesProcessed := queryStats.LoadSamplesProcessedPerStep()
+	if len(samplesProcessed) > 0 {
+		// TODO: reuse stats.StepStat proto in Extent, instead of defining duplicate
```
are you still doing this later?

Konstantinov-Innokentii replied:
Yep, it seems Marco found the solution, but I better do it after I merge this one.

dimitarvdimitrov commented on pkg/frontend/querymiddleware/split_and_cache.go:
```diff
@@ -304,6 +313,28 @@ func (s *splitAndCacheMiddleware) Do(ctx context.Context, req MetricsQueryReques
 		}
 	}
 
+	samplesProcessed := queryStats.LoadSamplesProcessedPerStep()
```
nit: can we tuck these new changes away in a function? `Do` as broad as it sounds, is already doing a lot

dimitarvdimitrov commented on pkg/frontend/transport/handler.go:
```diff
@@ -151,7 +151,8 @@ func NewHandler(cfg HandlerConfig, roundTripper http.RoundTripper, log log.Logge
 			Name: "cortex_query_samples_processed_cache_adjusted_total",
-			Help: "Number of samples processed to execute a query taking in account the original number of samples processed for parts of the query results looked up from the cache. Note, that this number is approximate.",
+			Help: "Number of samples processed to execute a query taking in account the original number of samples processed for parts of the query results looked up from the cache. " +
+				"Note, that it will be reported proprely only in conjunction with `-query-frontend.cache-samples-processed-stats` enabled. Otherwise, it will be 0.",
```
typo

dimitarvdimitrov replied:
also use present tense

dimitarvdimitrov commented on pkg/frontend/transport/handler.go:
```diff
@@ -329,20 +330,19 @@ func (f *Handler) reportQueryStats(
 	numChunks := stats.LoadFetchedChunks()
 	numIndexBytes := stats.LoadFetchedIndexBytes()
-	sharded := strconv.FormatBool(stats.GetShardedQueries() > 0)
+	sharded := strconv.FormatBool(stats.LoadShardedQueries() > 0)
```
nice catch, but changes like these are sometimes easier to review in a separate PR, especially when it's bundled with a lot of changes like this PR

Konstantinov-Innokentii replied:
Some tests were panicing without it, after I wrapped protogenerated stats into the "QueryStats".

dimitarvdimitrov commented on pkg/querier/stats/stats.go:
```diff
@@ -52,194 +58,213 @@ func (s *Stats) AddWallTime(t time.Duration) {
 }
 
+func (s *QueryStats) AddSamplesProcessedPerStep(points []StepStat) {
+	s.mx.Lock()
```
can we use `atomic.Pointer` here instead of a mutex? That way we don't need two structs and the slice is accessed in the same way as the rest of the fields

dimitarvdimitrov replied:
this works for me 


```go
func TestAtomicPointer(t *testing.T) {
	type testStruct struct {
		field []StepStat
	}

	strct := &testStruct{}
	newStats := []StepStat{{}, {}}
	atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&(strct.field))), unsafe.Pointer(&newStats))

	stats := (*[]StepStat)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&(strct.field)))))
	assert.Len(t, *stats, 2)
}
```

dimitarvdimitrov replied:
yeah it won't i wrote this comment before realziignwe should merge not overwrite

dimitarvdimitrov commented on pkg/querier/stats/stats_test.go:
```diff
@@ -170,6 +175,10 @@ func TestStats_Merge(t *testing.T) {
 		stats2.AddSplitQueries(11)
 		stats2.AddQueueTime(10 * time.Second)
 		stats2.AddSamplesProcessed(20)
+		stats2.AddSamplesProcessedPerStep([]StepStat{
```
in this test case it's not clear what we're testing  - that the stats are deduplicated or that we pick the first or the last or just any of them

dimitarvdimitrov commented on pkg/querier/stats/stats.go:
```diff
@@ -52,194 +58,213 @@ func (s *Stats) AddWallTime(t time.Duration) {
 	return atomic.LoadUint32(&s.SpunOffSubqueries)
 }
 
+func (s *QueryStats) AddSamplesProcessedPerStep(points []StepStat) {
```
shouldn't we be merging the points here? i think we share a `stats` instance for multiple queries, so the promql engine can invoke our `StatsRenderer` concurrently (that's why all fields are accessed atomically too i think)

Konstantinov-Innokentii replied:
Makes sense

Konstantinov-Innokentii commented on development/mimir-microservices-mode/docker-compose.jsonnet:
```diff
@@ -38,6 +38,9 @@ std.manifestYamlDoc({
     enable_query_tee: false,
+
+    // If true, enable per-step stats
+    enable_cached_samples_stats: true,
```
But this change it when you run mimir locally in a microservice mode, right?

Konstantinov-Innokentii replied:
Right, so I believe the change in [development/mimir-microservices-mode/config/mimir.yaml](https://github.com/grafana/mimir/pull/11719/files#diff-bd01f80874c5c272fa139e62b87aac503a98af5273c284b0f85ed8f86872bd76) is enough?

Konstantinov-Innokentii commented on pkg/frontend/querymiddleware/codec.go:
```diff
@@ -157,6 +160,14 @@ type MetricsQueryRequest interface {
+	// stats parameter controls prometheus query stats collection in the querier.
+	// Note, that these are not returned in the PrometheusResponse, but collected and used for querier.Stats.
+	// Possible values are:
+	// - any non-empty string: stats are collected
```
No, we don't. It affects only the PerStepStats collection and reporting. Even if someone provided "stats=all" it will not return anything.

Konstantinov-Innokentii commented on CHANGELOG.md:
```diff
@@ -84,6 +84,7 @@
   * `cortex_ingest_storage_writer_kafka_request_throttled_seconds`: measures how long Kafka requests have been throttled by the Kafka client.
 * [ENHANCEMENT] Distributor: Add per-user `cortex_distributor_sample_delay_seconds` to track delay of ingested samples with regard to wall clock. #11573
 * [ENHANCEMENT] Distributor: added circuit breaker to not produce Kafka records at all if the context is already canceled / expired. This applied only when experimental ingest storage is enabled. #11768
+* [ENHANCEMENT] Query-frontend: accurate tracking of samples processed from cache #11719
```
I'm happy to add it, but I'm unsure what you want me to mention there. From what I see, it's a API doc, but we didn't change user-facing apis?

dimitarvdimitrov replied:
oh fair point, this wasn't changed in this PR, but I still think it should be mentioned in the reference docs. especially its interaction with the per step caching flag

Konstantinov-Innokentii replied:
Sorry, but what's reference docs?

dimitarvdimitrov commented on integration/query_frontend_samples_processed_test.go:
```diff
@@ -0,0 +1,524 @@
+	"github.com/grafana/mimir/integration/e2emimir"
+)
+
+func TestQueryFrontendStatsFromResultsCacheShouldBeSameWhenWholeQueryIsCached(t *testing.T) {
```
i mean all tests. only `TestQueryFrontendStatsFromResultsCacheShouldBeSameWhenWholeQueryIsCached` took a minute. I think we can move these to unit tests where they'd take ~second without losing the strength of the assertions

Konstantinov-Innokentii commented:
> i left a few nitpicks, but also a few important comments like if we shouldn't be extending instead of overwriting the slice of stats in the querier stats.

Thanks!
> All of the nitpicks you can assume i'm ok to address in a follow-up PR or if it's a convo, we can continue after merging.

I believe I addressed most of them, except the one about integration tests, but we can continue the convo after merging the PR
> Did you have a chance to see the local dev setup instrutions I sent you yesterday? It would be nice to run this locally and see that the right metrics are incremented for sharded, split queries. You can even give it a go in a dev cell to see if it matches napkin math (like approximates the ratio between the stats for cached/uncached bytes in "query stats" logs - results_cache_miss_bytes and results_cache_hit_bytes)

Yep, I ran it locally and tested this. That's why I wrote integration tests, to not to keep doing it after every change :)

Konstantinov-Innokentii commented on integration/query_frontend_samples_processed_test.go:
```diff
@@ -0,0 +1,524 @@
+
+	testCases := []struct {
+		name                   string
+		splitQueriesByInterval string
```
Done, thanks

dimitarvdimitrov reviewed (approved):
nice one, I think we can merge it once you take care of the failing CI
